### DISCUSSION TITLE
App router misses authentication and  authorisation considerations for business apps

### DISCUSSION THREAD

**Discussion by slava-lu** [vercel/next.js#60725] [App Router]
There is an article https://nextjs.org/blog/security-nextjs-server-components-actions which covers how you can write a todo app.

But any more or less business useful application out there has 50+ APIs that has to be grouped somehow based on the user access rights. 

In Express.js world you can group routes with the Router and then apply some middleware to the group.  This approach scales endlessly. 

But how do I do it with Next.js? Now instead of the APIs we have just functions that access a database. How can I group my data access functions and make sure they return data only if my User object contains the necessary role?  I do not even know how to start thinking about it. 

Second issue. Before running a function I want to check if a user has been blocked by running some Select sql from the user table.  A trivial task in Espress.js and its middleware feature that looks to me to be impossible with Next.js if I have 50+ data access functions.

Did I miss anything from the new paradigm how server components and server actions work?

**Comment by 808vita** [comment]
[Doc link ](https://nextjs.org/docs/pages/building-your-application/routing/middleware) for applying middleware

**Convention**
Use the file middleware.ts (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.


`middleware.ts`
```
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
 
// This function can be marked `async` if using `await` inside
export function middleware(request: NextRequest) {
  return NextResponse.redirect(new URL('/home', request.url))
}
 

export const config = {
  matcher: '/about/:path*',
}
```



**Conditional Statements**

`middleware.ts`

```

import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
 
export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/about')) {
    return NextResponse.rewrite(new URL('/about-2', request.url))
  }
 
  if (request.nextUrl.pathname.startsWith('/dashboard')) {
    return NextResponse.rewrite(new URL('/dashboard/user', request.url))
  }
}
```


**Producing a Response**
You can respond from Middleware directly by returning a Response or NextResponse instance. (This is available since Next.js v13.1.0

`middleware.ts`

```

import { NextRequest } from 'next/server'
import { isAuthenticated } from '@lib/auth'
 
// Limit the middleware to paths starting with `/api/`
export const config = {
  matcher: '/api/:function*',
}
 
export function middleware(request: NextRequest) {
  // Call our authentication function to check the request
  if (!isAuthenticated(request)) {
    // Respond with JSON indicating an error message
    return Response.json(
      { success: false, message: 'authentication failed' },
      { status: 401 }
    )
  }
}
```

**Reply by slava-lu** [reply]
Sorry, but you missed the point a bit.  Next.js new React Server Component (RSC) concept does not use API handlers or route whatsoever. There is no /api/:function". There is no NextResponse or Response or Request. 
They tell us "You do need the API because you have RSC. Jut access database from RSC".  But they do NOT tell us how we write something bigger as todo app with RSC.

**Reply by 808vita** [reply]
I still think you would need to use middlewares.
___
[snippet from here ](https://nextjs.org/learn/dashboard-app/mutating-data)

_Good to know: In HTML, you'd pass a URL to the action attribute. This URL would be the destination where your form data should be submitted (usually an API endpoint).
However, in React, the action attribute is considered a special prop - meaning React builds on top of it to allow actions to be invoked.
**Behind the scenes, Server Actions create a POST API endpoint.** This is why you don't need to create API endpoints manually when using Server Actions._

[security-nextjs-server-components-actions#custom-routes-and-middleware](https://nextjs.org/blog/security-nextjs-server-components-actions#custom-routes-and-middleware)
Custom Routes and Middleware
Custom Route Handlers and Middleware are considered low level escape hatches for features that cannot be implemented using any other built-in functionality.
...
To make writing matchers easier Next.js App Router always uses the page's plain URL for both initial HTML, client navigations and Server Actions. Client navigations use ?_rsc=... search param as a cache breaker.

Server Actions live on the page they're used on and as such inherit the same access control. If Middleware allows reading a page, you can also invoke actions on that page. To limit access to Server Actions on a page, you can ban the POST HTTP method on that page.

___

Given that `server actions` automatically creates an end point and Next.js App Router always uses the page's plain URL for Server Actions- so it does make sense to use middlewares.
Additionally if required you can export server actions functions which takes a callback function and execute required actions if it passes required logic.

**Comment by laurent512** [comment]
I am also looking for the right approach to put an authorization layer before executing server actions. 
Currently, this verification is repeated for each server action #DRY...

Would love to be able to have a kind of file based logic e.g. if server actions is put in serveractions/logicA , authorizationLogicA will apply ;
